File: web/src/lib/api/models/app/advisor/Advisor.ts
// Advisor: continuously-evaluated recommendations about deliverability,
// mailbox configuration, warmup, campaign performance, copy, and list hygiene.
// Findings are surfaced where the fix lives (the campaign, the mailbox, the
// deliverability page) rather than in a separate inbox.

export type AdvisorSeverity = "critical" | "high" | "medium" | "low";

export type AdvisorCategory =
    | "deliverability"
    | "mailbox"
    | "warmup"
    | "campaign"
    | "copy"
    | "list";

// The dashboard nav tab a finding belongs to. Drives which tab shows a badge.
export type AdvisorSurface =
    | "campaigns"
    | "emails"
    | "deliverability"
    | "contacts"
    | "analytics"
    | "settings";

export type AdvisorStatus = "open" | "snoozed" | "dismissed" | "applied" | "resolved";

// One line of the before/after shown in the fix drawer. Every one-click fix
// renders its full effect this way before the user confirms.
export interface AdvisorPreviewChange {
    field: string;
    from: string;
    to: string;
}

export interface AdvisorAction {
    tool: string;
    args: unknown;
    label: string;
    // True when autopilot is allowed to apply this one unattended: a bounded,
    // reversible settings change that only moves in the safe direction.
    auto?: boolean;
    preview?: AdvisorPreviewChange[];
    undo?: { tool: string; args: unknown };
}

// What an agent fix reports back.
export interface AdvisorAgentResult {
    finding_id: string;
    // False when the agent read the current state and decided nothing needed
    // changing, which leaves the finding open.
    applied: boolean;
    summary: string;
    // The tools it called, in order. The checkable part of the receipt.
    steps?: string[];
}

export interface AdvisorFinding {
    id: string;
    organization_id: string;
    detector_key: string;
    category: AdvisorCategory;
    severity: AdvisorSeverity;
    surface: AdvisorSurface;

    entity_type?: string;
    entity_id?: string;
    entity_label?: string;

    // The entity this finding belongs to when it differs from the subject: a
    // step's copy problem belongs to its campaign, and the campaign row is
    // where someone goes looking for it.
    parent_type?: string;
    parent_id?: string;

    status: AdvisorStatus;
    impact: number;

    title: string;
    // How this finding names itself when several of its kind are shown
    // together, with a {count} placeholder. Empty means it always stands alone.
    group_title?: string;
    detail: string;
    remedy: string;
    // The ordered manual steps, set only by checks with no one-click fix. When
    // empty the card falls back to the remedy prose.
    steps?: string[];
    // Whether an agent can resolve this by editing something the platform owns.
    // False for anything living outside it, like a DNS record, so those show
    // their steps rather than a button that cannot succeed.
    agent_fixable?: boolean;
    // False while the card carries the built-in copy rather than AI-written
    // copy. The card is fully usable either way.
    narrated: boolean;

    // The numbers the detector fired on, rendered as chips under the detail.
    evidence?: Record<string, unknown>;
    action?: AdvisorAction;

    first_seen_at: string;
    last_seen_at: string;
    snoozed_until?: string;
    applied_at?: string;
    applied_result?: string;
}

export interface AdvisorSurfaceCount {
    surface: AdvisorSurface;
    total: number;
    critical: number;
    high: number;
}

export interface AdvisorSummary {
    score: number;
    total: number;
    critical: number;
    high: number;
    medium: number;
    low: number;
    surfaces: AdvisorSurfaceCount[];
    last_run_at?: string;
}

export interface AdvisorSettings {
    organization_id: string;
    enabled: boolean;
    muted_categories: string[];
    muted_detectors: string[];
    min_severity: AdvisorSeverity;
    // Applies the auto-safe fixes without asking, as the member who switched
    // it on.
    autopilot: boolean;
    autopilot_actor_id?: string;
    updated_at: string;
}

export interface AdvisorFindingsQuery {
    surface?: AdvisorSurface;
    category?: AdvisorCategory;
    entityType?: string;
    entityId?: string;
    status?: AdvisorStatus[];
    limit?: number;
}

// Presentation tokens, kept beside the types so every advisor surface renders
// severity identically.
export const SEVERITY_LABEL: Record<AdvisorSeverity, string> = {
    critical: "Critical",
    high: "Needs attention",
    medium: "Worth fixing",
    low: "Suggestion",
};

// Tints rather than fills. A translucent wash of the severity colour lets
// whatever is underneath (a row hover, the drawer's frosted panel) show
// through, so a chip reads as a mark on the surface instead of a sticker
// sitting on top of it.
export const SEVERITY_CHIP: Record<AdvisorSeverity, string> = {
    critical: "bg-rose-500/10 text-rose-700 border-rose-500/20",
    high: "bg-orange-500/10 text-orange-700 border-orange-500/20",
    medium: "bg-amber-500/10 text-amber-700 border-amber-500/25",
    low: "bg-sky-500/10 text-sky-700 border-sky-500/20",
};

export const SEVERITY_DOT: Record<AdvisorSeverity, string> = {
    critical: "bg-rose-500",
    high: "bg-orange-500",
    medium: "bg-amber-500",
    low: "bg-sky-500",
};

export const SEVERITY_RANK: Record<AdvisorSeverity, number> = {
    critical: 4,
    high: 3,
    medium: 2,
    low: 1,
};

// Tone for the inline indicator that sits on the row the problem is about. It
// has to read as a status on someone else's row rather than a control of its
// own, so it stays borderless until hovered.
export const SEVERITY_ROW: Record<AdvisorSeverity, string> = {
    critical: "bg-rose-500/10 text-rose-700 hover:bg-rose-500/20",
    high: "bg-orange-500/10 text-orange-700 hover:bg-orange-500/20",
    medium: "bg-amber-500/10 text-amber-700 hover:bg-amber-500/25",
    low: "bg-sky-500/10 text-sky-700 hover:bg-sky-500/20",
};

// The one-word verdict the row indicator shows. Deliberately shorter than
// SEVERITY_LABEL: a table cell has room for a word, not a phrase.
export const SEVERITY_SHORT: Record<AdvisorSeverity, string> = {
    critical: "Urgent",
    high: "Fix",
    medium: "Tune",
    low: "Tip",
};

export const CATEGORY_LABEL: Record<AdvisorCategory, string> = {
    deliverability: "Deliverability",
    mailbox: "Mailbox",
    warmup: "Warmup",
    campaign: "Campaign",
    copy: "Copy",
    list: "List",
};

// A run of findings from the same check, collapsed into one card. One
// misconfiguration repeated across twenty mailboxes should read as one problem
// with twenty subjects, not twenty problems.
export interface AdvisorGroup {
    key: string;
    // lead is the most urgent member, and supplies the shared detail and remedy.
    lead: AdvisorFinding;
    members: AdvisorFinding[];
}

// groupFindings collapses runs of the same check into one entry. A check only
// collapses if it declared a group title, and only from three members up: two
// cards naming two specific mailboxes are more useful than one card saying
// "2 mailboxes".
export function groupFindings(findings: AdvisorFinding[], minSize = 3): AdvisorGroup[] {
    const byKey = new Map<string, AdvisorFinding[]>();
    for (const f of findings) {
        const list = byKey.get(f.detector_key);
        if (list) list.push(f);
        else byKey.set(f.detector_key, [f]);
    }

    const out: AdvisorGroup[] = [];
    for (const [key, members] of byKey) {
        const collapse = members.length >= minSize && Boolean(members[0].group_title);
        if (collapse) {
            out.push({ key, lead: members[0], members });
        } else {
            for (const f of members) out.push({ key: f.id, lead: f, members: [f] });
        }
    }

    // Re-sort: a collapsed group takes the urgency of its most urgent member,
    // so a run of criticals never sinks below a single medium.
    return out.sort((a, b) => {
        const rank = SEVERITY_RANK[b.lead.severity] - SEVERITY_RANK[a.lead.severity];
        if (rank !== 0) return rank;
        if (b.members.length !== a.members.length) return b.members.length - a.members.length;
        return b.lead.impact - a.lead.impact;
    });
}

// groupTitle renders a group's heading, falling back to the lead's own title
// for a check that never declared one.
export function groupTitle(group: AdvisorGroup): string {
    if (group.members.length < 2 || !group.lead.group_title) return group.lead.title;
    return group.lead.group_title.replace("{count}", String(group.members.length));
}

// resolutionSteps is the ordered how-to for a finding with no one-click fix.
// Empty means the remedy prose is the whole answer.
export function resolutionSteps(finding: AdvisorFinding): string[] {
    return finding.steps ?? [];
}

// findingLink is the way to the screen where a manual fix is actually made. A
// step's problem sends you to its campaign, since a step has no page of its own.
export function findingLink(finding: AdvisorFinding): { href: string; label: string } | null {
    if (finding.entity_type === "campaign" && finding.entity_id) {
        return { href: `/app/campaigns/${finding.entity_id}`, label: "Open the campaign" };
    }
    if (finding.entity_type === "step" && finding.parent_id) {
        return { href: `/app/campaigns/${finding.parent_id}`, label: "Open the campaign" };
    }
    if (finding.entity_type === "email_account" && finding.entity_id) {
        return { href: `/app/emails?mailbox=${finding.entity_id}`, label: "Open the mailbox" };
    }
    switch (finding.surface) {
        case "deliverability":
            return { href: "/app/deliverability", label: "Open deliverability" };
        case "contacts":
            return { href: "/app/contacts", label: "Open contacts" };
        case "settings":
            return { href: "/app/settings/workspace", label: "Open workspace settings" };
        default:
            return null;
    }
}

// worstSeverity is the tone a group of findings should take: the most urgent
// one present. A row with a critical and three suggestions is a critical row.
export function worstSeverity(findings: AdvisorFinding[]): AdvisorSeverity | null {
    let worst: AdvisorSeverity | null = null;
    for (const f of findings) {
        if (!worst || SEVERITY_RANK[f.severity] > SEVERITY_RANK[worst]) worst = f.severity;
    }
    return worst;
}

// sortFindings orders a list the way every advisor surface shows it: most
// urgent first, applied ones last so a just-fixed card does not jump to the top.
export function sortFindings(findings: AdvisorFinding[]): AdvisorFinding[] {
    return [...findings].sort((a, b) => {
        if ((a.status === "applied") !== (b.status === "applied")) {
            return a.status === "applied" ? 1 : -1;
        }
        const rank = SEVERITY_RANK[b.severity] - SEVERITY_RANK[a.severity];
        return rank !== 0 ? rank : b.impact - a.impact;
    });
}

// An advisor run sliced by the entity each finding is about, so a list page can
// fetch its surface once and hand every row its own advice.
export interface AdvisorEntityIndex {
    // get returns the findings attached to one entity, most urgent first.
    get(entityId?: string | null): AdvisorFinding[];
    // unattached are the findings no row can carry: workspace-wide settings, or
    // an entity that is not on this page. The page-level summary shows these.
    unattached: AdvisorFinding[];
    // total counts every finding in the run, attached or not.
    total: number;
}

const EMPTY: AdvisorFinding[] = [];

// indexByEntity buckets a run by subject AND by parent, because the row a
// finding belongs on is not always the row it is about: a step's copy problem
// has no row of its own and belongs on its campaign.
export function indexByEntity(findings: AdvisorFinding[]): AdvisorEntityIndex {
    const byEntity = new Map<string, AdvisorFinding[]>();
    const unattached: AdvisorFinding[] = [];

    const push = (id: string, f: AdvisorFinding) => {
        const list = byEntity.get(id);
        if (list) {
            // A finding whose parent is also its subject must not be listed twice.
            if (!list.some((x) => x.id === f.id)) list.push(f);
        } else {
            byEntity.set(id, [f]);
        }
    };

    for (const f of findings) {
        if (f.entity_id) push(f.entity_id, f);
        if (f.parent_id) push(f.parent_id, f);
        if (!f.entity_id && !f.parent_id) unattached.push(f);
    }

    for (const [id, list] of byEntity) byEntity.set(id, sortFindings(list));

    return {
        get: (entityId) => (entityId ? (byEntity.get(entityId) ?? EMPTY) : EMPTY),
        unattached: sortFindings(unattached),
        total: findings.length,
    };
}

// Evidence keys are snake_case machine names; this renders them as readable
// labels without needing a per-detector translation table.
export function evidenceLabel(key: string): string {
    return key
        .replace(/_percent$/, " (%)")
        .replace(/_/g, " ")
        .replace(/\b7d\b/, "last 7 days")
        .replace(/\b30d\b/, "last 30 days")
        .replace(/^./, (c) => c.toUpperCase());
}

export function evidenceValue(value: unknown): string {
    if (value === null || value === undefined) return "—";
    if (typeof value === "boolean") return value ? "yes" : "no";
    if (Array.isArray(value)) return value.map((v) => String(v)).join(", ");
    if (typeof value === "number") return Number.isInteger(value) ? String(value) : value.toFixed(2);
    return String(value);
}

File: internal/models/advisor.go
package models

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"math"
	"time"

	"github.com/google/uuid"
)

// AdvisorSeverity ranks how urgently a finding needs attention. It drives
// ordering, the nav-tab badge tone, and whether a new finding is worth a toast.
type AdvisorSeverity string

const (
	// AdvisorCritical means reputation or delivery is actively being damaged
	// right now (complaint/bounce rates in the documented hard-block bands, a
	// mailbox sending cold mail while quarantined from the warmup pool).
	AdvisorCritical AdvisorSeverity = "critical"
	// AdvisorHigh means a real problem that will become critical if ignored.
	AdvisorHigh AdvisorSeverity = "high"
	// AdvisorMedium means a meaningful improvement with no immediate risk.
	AdvisorMedium AdvisorSeverity = "medium"
	// AdvisorLow means a polish-level suggestion.
	AdvisorLow AdvisorSeverity = "low"
)

// severityRank orders severities most-urgent-first for sorting and threshold
// comparisons.
var severityRank = map[AdvisorSeverity]int{
	AdvisorCritical: 4,
	AdvisorHigh:     3,
	AdvisorMedium:   2,
	AdvisorLow:      1,
}

// Rank returns the numeric urgency of a severity (higher is more urgent, 0 for
// an unknown value).
func (s AdvisorSeverity) Rank() int { return severityRank[s] }

// AtLeast reports whether s is at least as urgent as min.
func (s AdvisorSeverity) AtLeast(min AdvisorSeverity) bool { return s.Rank() >= min.Rank() }

// AdvisorCategory groups findings by the kind of problem they describe. Orgs
// can mute a whole category.
type AdvisorCategory string

const (
	AdvisorCategoryDeliverability AdvisorCategory = "deliverability"
	AdvisorCategoryMailbox        AdvisorCategory = "mailbox"
	AdvisorCategoryWarmup         AdvisorCategory = "warmup"
	AdvisorCategoryCampaign       AdvisorCategory = "campaign"
	AdvisorCategoryCopy           AdvisorCategory = "copy"
	AdvisorCategoryList           AdvisorCategory = "list"
)

// AdvisorSurface is the dashboard nav tab a finding belongs to. The badge count
// and the inline strip both key off it, so advice always appears where the fix
// is made rather than in a separate inbox.
type AdvisorSurface string

const (
	AdvisorSurfaceCampaigns      AdvisorSurface = "campaigns"
	AdvisorSurfaceMailboxes      AdvisorSurface = "emails"
	AdvisorSurfaceDeliverability AdvisorSurface = "deliverability"
	AdvisorSurfaceContacts       AdvisorSurface = "contacts"
	AdvisorSurfaceAnalytics      AdvisorSurface = "analytics"
	AdvisorSurfaceSettings       AdvisorSurface = "settings"
)

// AdvisorStatus is a finding's lifecycle state.
type AdvisorStatus string

const (
	// AdvisorStatusOpen is an active, unaddressed finding.
	AdvisorStatusOpen AdvisorStatus = "open"
	// AdvisorStatusSnoozed hides the finding until SnoozedUntil passes. The
	// detector keeps re-confirming it in the background.
	AdvisorStatusSnoozed AdvisorStatus = "snoozed"
	// AdvisorStatusDismissed means a member said this isn't a problem for them.
	// It stays dismissed until the underlying condition clears and recurs.
	AdvisorStatusDismissed AdvisorStatus = "dismissed"
	// AdvisorStatusApplied means the one-click fix ran. It stays in this state
	// until the next evaluation confirms the condition is gone (-> resolved) or
	// still present (-> reopened).
	AdvisorStatusApplied AdvisorStatus = "applied"
	// AdvisorStatusResolved means the detector no longer fires.
	AdvisorStatusResolved AdvisorStatus = "resolved"
)

// AdvisorPreviewChange is one line of the before/after preview shown in the fix
// drawer. Every one-click fix renders its full effect this way before the user
// confirms; nothing is ever applied sight-unseen.
type AdvisorPreviewChange struct {
	Field string `json:"field"`
	From  string `json:"from"`
	To    string `json:"to"`
}

// AdvisorAction is the one-click remedy attached to a finding. It executes
// through the shared AI tool registry, which runs the tool AS the invoking user
// with their permission bits enforced, so the Advisor can never apply a change
// the member could not have made by hand.
type AdvisorAction struct {
	// Tool is the aitools registry name (e.g. "update_mailbox").
	Tool string `json:"tool"`
	// Args is the tool's JSON payload, fully materialized at detection time.
	Args json.RawMessage `json:"args"`
	// Label is the button text ("Lower the daily cap to 35").
	Label string `json:"label"`
	// Auto marks a fix that autopilot may apply unattended. It is true only for
	// a bounded settings change that moves in the safe direction and can be
	// undone: lowering a cap, widening a gap, resuming warmup. Anything that
	// stops sending, edits copy, or cannot be reverted stays false and waits
	// for a person, however obvious the fix looks.
	Auto bool `json:"auto,omitempty"`
	// Preview is the exact before/after the drawer renders.
	Preview []AdvisorPreviewChange `json:"preview,omitempty"`
	// Undo, when set, is the tool call that reverts this one. Surfaced as
	// "Undo" on the applied card.
	Undo *AdvisorUndo `json:"undo,omitempty"`
}

// AdvisorUndo reverts an applied action.
type AdvisorUndo struct {
	Tool string          `json:"tool"`
	Args json.RawMessage `json:"args"`
}

// AdvisorFinding is one piece of advice about one entity.
type AdvisorFinding struct {
	ID             uuid.UUID `json:"id"`
	OrganizationID uuid.UUID `json:"organization_id"`

	Fingerprint string          `json:"-"`
	DetectorKey string          `json:"detector_key"`
	Category    AdvisorCategory `json:"category"`
	Severity    AdvisorSeverity `json:"severity"`
	Surface     AdvisorSurface  `json:"surface"`

	EntityType  string     `json:"entity_type,omitempty"`
	EntityID    *uuid.UUID `json:"entity_id,omitempty"`
	EntityLabel string     `json:"entity_label,omitempty"`

	// ParentType / ParentID name the entity this finding belongs to when it
	// differs from the subject: a step's copy problem belongs to its campaign,
	// and the campaign page is where someone goes looking for it.
	ParentType string     `json:"parent_type,omitempty"`
	ParentID   *uuid.UUID `json:"parent_id,omitempty"`

	Status AdvisorStatus `json:"status"`
	Impact int           `json:"impact"`

	Title string `json:"title"`
	// GroupTitle names the finding when several of its kind are listed
	// together, with a {count} placeholder ("{count} mailboxes are capped above
	// the safe band"). Empty means this finding is always shown on its own.
	GroupTitle string `json:"group_title,omitempty"`
	Detail     string `json:"detail"`
	Remedy     string `json:"remedy"`
	// Steps is the ordered manual how-to, set only by checks with no one-click
	// fix. Empty means the remedy prose is the whole answer.
	Steps []string `json:"steps,omitempty"`
	// AgentFixable is true when an agent can resolve this by editing something
	// the platform owns. False for anything that lives outside it, like a DNS
	// record, so the client shows the steps rather than a button that cannot
	// succeed. Computed per request, never stored.
	AgentFixable bool `json:"agent_fixable"`
	// Narrated is false while the card still shows the built-in fallback copy
	// (AI unconfigured, or narration not run yet). The card is fully usable
	// either way; this only tells the client whether to offer "rewrite".
	Narrated bool `json:"narrated"`

	Evidence json.RawMessage `json:"evidence,omitempty"`
	Action   *AdvisorAction  `json:"action,omitempty"`

	FirstSeenAt time.Time  `json:"first_seen_at"`
	LastSeenAt  time.Time  `json:"last_seen_at"`
	ResolvedAt  *time.Time `json:"resolved_at,omitempty"`

	SnoozedUntil  *time.Time `json:"snoozed_until,omitempty"`
	DismissedAt   *time.Time `json:"dismissed_at,omitempty"`
	DismissReason string     `json:"dismiss_reason,omitempty"`

	AppliedAt     *time.Time `json:"applied_at,omitempty"`
	AppliedBy     *uuid.UUID `json:"applied_by,omitempty"`
	AppliedResult string     `json:"applied_result,omitempty"`
}

// EvidenceHash fingerprints the finding's evidence so a re-run can tell
// "same problem, same numbers" (keep the narration) from "same problem, the
// numbers moved" (re-narrate). Detectors round the values they put in Evidence
// into bands, so ordinary drift does not churn the copy.
func (f *AdvisorFinding) EvidenceHash() string {
	sum := sha256.Sum256(f.Evidence)
	return hex.EncodeToString(sum[:8])
}

// AdvisorSurfaceCount is the per-nav-tab badge payload.
type AdvisorSurfaceCount struct {
	Surface AdvisorSurface `json:"surface"`
	// Total is every open finding on this surface; Critical/High drive the
	// badge tone so a tab never screams about a low-severity nit.
	Total    int `json:"total"`
	Critical int `json:"critical"`
	High     int `json:"high"`
}

// AdvisorSummary is the org-wide rollup: what the badges show, plus a single
// health score so the trend is legible at a glance.
type AdvisorSummary struct {
	// Score is 100 minus the weighted cost of every open finding, floored at 0.
	Score    int                   `json:"score"`
	Total    int                   `json:"total"`
	Critical int                   `json:"critical"`
	High     int                   `json:"high"`
	Medium   int                   `json:"medium"`
	Low      int                   `json:"low"`
	Surfaces []AdvisorSurfaceCount `json:"surfaces"`
	// LastRunAt is when the engine last evaluated this org (nil before the
	// first run).
	LastRunAt *time.Time `json:"last_run_at,omitempty"`
}

// advisorSeverityCost is how much each open finding weighs against the org's
// Advisor score. Weighted so one critical finding matters more than a dozen
// low-severity nits, which is the ordering a sender actually cares about.
var advisorSeverityCost = map[AdvisorSeverity]float64{
	AdvisorCritical: 25,
	AdvisorHigh:     10,
	AdvisorMedium:   4,
	AdvisorLow:      1,
}

// advisorScoreScale tunes how fast the score falls. A single critical finding
// lands around 78; the curve then flattens, so the twentieth medium-severity
// suggestion moves the number far less than the first critical one did.
const advisorScoreScale = 100.0

// AdvisorScore folds open-finding counts into a 0-100 health score.
//
// The penalty saturates rather than accumulating linearly. A linear score hits
// zero at around four high-severity findings, after which a workspace with four
// problems and a workspace with forty look identical, and fixing three of them
// changes nothing on screen. An exponential decay keeps the number honest
// across the range a workspace actually lives in: it never reaches zero, and
// every fix moves it until the tail flattens out.
func AdvisorScore(critical, high, medium, low int) int {
	cost := float64(critical)*advisorSeverityCost[AdvisorCritical] +
		float64(high)*advisorSeverityCost[AdvisorHigh] +
		float64(medium)*advisorSeverityCost[AdvisorMedium] +
		float64(low)*advisorSeverityCost[AdvisorLow]
	if cost <= 0 {
		return 100
	}
	score := int(math.Round(100 * math.Exp(-cost/advisorScoreScale)))
	if score < 1 {
		return 1
	}
	return score
}

// AdvisorSettings holds the org's advisor controls.
type AdvisorSettings struct {
	OrganizationID  uuid.UUID       `json:"organization_id"`
	Enabled         bool            `json:"enabled"`
	MutedCategories []string        `json:"muted_categories"`
	MutedDetectors  []string        `json:"muted_detectors"`
	MinSeverity     AdvisorSeverity `json:"min_severity"`
	// Autopilot applies auto-safe fixes on its own. Off by default.
	Autopilot bool `json:"autopilot"`
	// AutopilotActorID is the member autopilot acts as. Set automatically to
	// whoever switches it on; autopilot stops if they leave the org.
	AutopilotActorID *uuid.UUID `json:"autopilot_actor_id,omitempty"`
	UpdatedAt        time.Time  `json:"updated_at"`
}

// AutopilotMaxPerRun bounds how many fixes autopilot may apply in one
// evaluation. A misconfiguration should cost a handful of reversible changes,
// not a whole workspace rewritten while nobody was looking.
const AutopilotMaxPerRun = 10

// DefaultAdvisorSettings is what an org gets before it has ever written a
// settings row: everything on, nothing muted.
func DefaultAdvisorSettings(orgID uuid.UUID) *AdvisorSettings {
	return &AdvisorSettings{
		OrganizationID:  orgID,
		Enabled:         true,
		MutedCategories: []string{},
		MutedDetectors:  []string{},
		MinSeverity:     AdvisorLow,
	}
}

// AdvisorAgentResult is what an agent fix reports back.
type AdvisorAgentResult struct {
	FindingID uuid.UUID `json:"finding_id"`
	// Applied is true only when the agent called a tool that changed something.
	// A run that read the campaign and decided nothing was wrong reports false,
	// so the finding stays open.
	Applied bool `json:"applied"`
	// Summary is the agent's own account of what it did.
	Summary string `json:"summary"`
	// Steps are the tool calls it made, in order. This is the part the member
	// can check against the audit log.
	Steps []string `json:"steps,omitempty"`
}

// AdvisorFeedback is a member's verdict on one finding.
type AdvisorFeedback struct {
	FindingID uuid.UUID `json:"finding_id"`
	Helpful   bool      `json:"helpful"`
	Reason    string    `json:"reason,omitempty"`
}

// AdvisorSnoozeRequest snoozes a finding for a bounded window.
type AdvisorSnoozeRequest struct {
	// Days must be 1-90. Anything else is rejected: an unbounded snooze is a
	// dismissal wearing a disguise, and the two should stay distinguishable.
	Days int `json:"days" binding:"required,min=1,max=90"`
}

// AdvisorDismissRequest records why a member rejected the advice. The reason is
// optional but feeds detector tuning.
type AdvisorDismissRequest struct {
	Reason string `json:"reason"`
}
